### ISSUE TITLE
Run static analyzer when building iOS engine dylib

### ISSUE THREAD

**Issue by jmagman** [flutter/flutter#91857]
We should be running static analysis errors when building iOS engine.  This will cause the build time to increase quite a lot, which should be evaluated.

I believe this can be done by passing `--clang-static-analyzer` into `gn` in the builder:
https://github.com/flutter/engine/blob/217d545631f9010d10f6f250422b6d18ba2b331c/ci/builders/mac_ios_engine.json#L31-L36
https://github.com/flutter/engine/blob/945a524bfc5c5eef25487ece2151db644d0eb256/tools/gn#L440

**Comment by jmagman** [comment]
@zanderso do you know where `clang_static_analyzer_wrapper.py` lives?

**Comment by jmagman** [comment]
Split the python problem out to https://github.com/flutter/flutter/issues/91860.

**Comment by zanderso** [comment]
The `clang_static_analyzer_wrapper.py` script was inherited from the chromium buildroot many eons ago. Chromium has since [abandoned it](https://chromium-review.googlesource.com/c/chromium/src/+/1538987) in favor of `clang-tidy`.

Unfortunately, I wasn't able to determine from the bugs linked from that CL which specific `clang-tidy` checks replace the functionality of the `clang_static_analyzer_wrapper.py` script.

@jmagman Do you know what flags we need to pass to `clang` or `clang-tidy` to do the static analysis we need? If so, we should probably just roll them into our build system directly instead of trying to get a deprecated chromium script to work.

**Comment by zanderso** [comment]
That is, I guess, do you know which of the flags from https://github.com/flutter/flutter/issues/91857#issuecomment-943828859 are the ones we need to add?

**Comment by zanderso** [comment]
I added these flags to objc compiles:

<details>
  <summary>Flags</summary>

```
      defines += [ "__clang_analyzer__" ]
      cflags_objcc = [
        "--analyze",
        "-fdiagnostics-show-note-include-stack",
        "-Xclang",
        "-analyzer-output=plist-multi-file",
        "-Xclang",
        "-analyzer-config",
        "-Xclang",
        "path-diagnostics-alternate=true",
        "-Xclang",
        "-analyzer-config",
        "-Xclang",
        "report-in-main-source-file=true",
        "-Xclang",
        "-analyzer-config",
        "-Xclang",
        "osx.NumberObjectConversion:Pedantic=true",
        "-Xclang",
        "-analyzer-checker",
        "-Xclang",
        "security.insecureAPI.UncheckedReturn",
        "-Xclang",
        "-analyzer-checker",
        "-Xclang",
        "security.insecureAPI.getpw",
        "-Xclang",
        "-analyzer-checker",
        "-Xclang",
        "security.insecureAPI.gets",
        "-Xclang",
        "-analyzer-checker",
        "-Xclang",
        "security.insecureAPI.mkstemp",
        "-Xclang",
        "-analyzer-checker",
        "-Xclang",
        "security.insecureAPI.mktemp",
        "-Xclang",
        "-analyzer-disable-checker",
        "-Xclang",
        "security.insecureAPI.rand",
        "-Xclang",
        "-analyzer-disable-checker",
        "-Xclang",
        "security.insecureAPI.strcpy",
        "-Xclang",
        "-analyzer-checker",
        "-Xclang",
        "security.insecureAPI.vfork",
      ]
```

</details>

but hit a linker error that I haven't had a chance to investigate yet:

```
ld: warning: ignoring file obj/third_party/swiftshader/src/OpenGL/libEGL/libEGL.OSXUtils.o, building for macOS-x86_64 but attempting to link with file built for unknown-unsupported file format ( 0x3C 0x3F 0x78 0x6D 0x6C 0x20 0x76 0x65 0x72 0x73 0x69 0x6F 0x6E 0x3D 0x22 0x31 )
Undefined symbols for architecture x86_64:
  "sw::OSX::GetNativeWindowSize(void*, int&, int&)", referenced from:
      egl::WindowSurface::initialize() in libEGL.Surface.o
      egl::WindowSurface::swap() in libEGL.Surface.o
  "sw::OSX::IsValidWindow(void*)", referenced from:
      egl::Display::isValidWindow(void*) in libEGL.Display.o
ld: symbol(s) not found for architecture x86_64
```

**Comment by jmagman** [comment]
I'm particularly interested in the Objective-C MRC warnings, which I think we can get with just `--analyze` and none of the other `-Xclang -analyzer-checker` flags (although if we can throw in any others listed we think are valuable that would be great).

Just tried it in a sample app:
```
main.m:12:3: warning: Incorrect decrement of the reference count of an object that is not owned at this point by the caller [osx.cocoa.RetainCount]
  [array release];
  ^~~~~~~~~~~~~~~
1 warning generated.
```

**Comment by zanderso** [comment]
@jmagman Does `clang-tidy` have any of the checks you're interested in? https://clang.llvm.org/extra/clang-tidy/checks/list.html. If so, we might be able to just add them to the Engine's `.clang-tidy` file: https://github.com/flutter/engine/blob/master/.clang-tidy

**Comment by jmagman** [comment]
Interesting, it already has them on.
```
../buildtools/mac-x64/clang/bin/clang-tidy  --list-checks
...
    clang-analyzer-osx.cocoa.Dealloc
    clang-analyzer-osx.cocoa.IncompatibleMethodTypes
    clang-analyzer-osx.cocoa.Loops
    clang-analyzer-osx.cocoa.MissingSuperCall
    clang-analyzer-osx.cocoa.NSAutoreleasePool
    clang-analyzer-osx.cocoa.NSError
    clang-analyzer-osx.cocoa.NilArg
    clang-analyzer-osx.cocoa.NonNilReturnValue
    clang-analyzer-osx.cocoa.ObjCGenerics
    clang-analyzer-osx.cocoa.RetainCount
    clang-analyzer-osx.cocoa.RetainCountBase
...
```
It's only running for `host_debug/compile_commands.json`, which runs on the macOS shell but not iOS
https://github.com/flutter/engine/blob/1ae721c0230e8e3f3618f7e9d301ad2133b2fd61/ci/lint.sh#L35

However, when I change that to an `ios_` out directory, it still doesn't run the analyzer.  To test this, I changed
https://github.com/flutter/engine/blob/1ae721c0230e8e3f3618f7e9d301ad2133b2fd61/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm#L469
to the incorrectly overreleased code:
```objc
    _index = index;

    NSArray *array = @[@"test"];
    [array release];
    [array release];
    [array release];
```

The file gets picked up, but no lint problems found:
```
🔶 linting flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm
Jobs: 100% done,   1/1   completed,  0 in progress,   0 pending,   0 failed.    

No lint problems found.
```

When I run `clang-tidy` directly on the file, I see `clang-analyzer` warnings, but not `google-` warnings:
```
$ ../../buildtools/mac-x64/clang/bin/clang-tidy /Users/m/Projects/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm --config= --  ++ -MD  obj/flutter/shell/platform/darwin/ios/framework/Source/flutter_framework_source.FlutterTextInputPlugin.o.d  -DFLUTTER_FRAMEWORK=1 -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D_FORTIFY_SOURCE=2 -D_LIBCPP_DISABLE_VISIBILITY_ANNOTATIONS -D_LIBCPP_ENABLE_THREAD_SAFETY_ANNOTATIONS -D_DEBUG -DSHELL_ENABLE_SOFTWARE -DSHELL_ENABLE_GL -DSHELL_ENABLE_METAL -DSK_GL -DSK_METAL -DSK_ENABLE_DUMP_GPU -DSK_CODEC_DECODES_JPEG -DSK_ENCODE_JPEG -DSK_CODEC_DECODES_PNG -DSK_ENCODE_PNG -DSK_CODEC_DECODES_WEBP -DSK_ENCODE_WEBP -DSK_HAS_WUFFS_LIBRARY -DSK_DISABLE_AAA -DSK_PARAGRAPH_LIBTXT_SPACES_RESOLUTION -DSK_LEGACY_INNER_JOINS -DSK_DISABLE_LEGACY_SHADERCONTEXT -DSK_DISABLE_LOWP_RASTER_PIPELINE -DSK_FORCE_RASTER_PIPELINE_BLITTER -DSK_DISABLE_EFFECT_DESERIALIZATION -DSK_ENABLE_SKSL -DSK_ASSUME_GL_ES=1 -DSK_ENABLE_API_AVAILABLE -DFLUTTER_RUNTIME_MODE_DEBUG=1 -DFLUTTER_RUNTIME_MODE_PROFILE=2 -DFLUTTER_RUNTIME_MODE_RELEASE=3 -DFLUTTER_RUNTIME_MODE_JIT_RELEASE=4 "-DDART_LEGACY_API=[[deprecated]]" -DFLUTTER_RUNTIME_MODE=1 -DFLUTTER_JIT_RUNTIME=1 -DU_USING_ICU_NAMESPACE=0 -DU_ENABLE_DYLOAD=0 -DUSE_CHROMIUM_ICU=1 -DU_ENABLE_TRACING=1 -DU_ENABLE_RESOURCE_TRACING=0 -DU_STATIC_IMPLEMENTATION -DICU_UTIL_DATA_IMPL=ICU_UTIL_DATA_FILE -DRAPIDJSON_HAS_STDSTRING -DRAPIDJSON_HAS_CXX11_RANGE_FOR -DRAPIDJSON_HAS_CXX11_RVALUE_REFS -DRAPIDJSON_HAS_CXX11_TYPETRAITS -DRAPIDJSON_HAS_CXX11_NOEXCEPT -DFLUTTER_API_SYMBOL_PREFIX=Embedder -DFLUTTER_NO_EXPORT  -I../.. -Igen -I../../third_party/skia -I../../third_party/vulkan/src -I../../third_party/vulkan/include -I../../flutter/third_party/txt/src -I../../third_party/harfbuzz/src -I../../third_party/icu/source/common -I../../third_party/icu/source/i18n -I../../third_party/rapidjson -I../../third_party/rapidjson/include -I../../third_party/dart/runtime -I../../third_party/dart/runtime/include -I../../flutter/third_party -I../../flutter -I../../flutter/shell/platform/darwin/common/framework/Headers -isysroot /Users/m/Applications/Xcode-13_rc.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator15.0.sdk -mios-simulator-version-min=9.0  -fno-strict-aliasing -arch x86_64 -fcolor-diagnostics -Wall -Wextra -Wendif-labels -Werror -Wno-missing-field-initializers -Wno-unused-parameter -Wno-unused-but-set-parameter -Wno-unused-but-set-variable -Wno-implicit-int-float-conversion -Wno-c99-designator -Wno-deprecated-copy -Wno-psabi -Wno-non-c-typedef-for-linkage -Wno-range-loop-construct -Wunguarded-availability -Wno-deprecated-declarations -fvisibility=hidden -stdlib=libc++ -Wstring-conversion -Wnewline-eof -O0 -g2   -Werror=overriding-method-mismatch -Werror=undeclared-selector -fvisibility-inlines-hidden -fobjc-call-cxx-cdtors -std=c++17 -fno-rtti -fno-exceptions -nostdinc++ -isystem /Users/m/Applications/Xcode-13_rc.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1  -c ../../flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm -o  obj/flutter/shell/platform/darwin/ios/framework/Source/flutter_framework_source.FlutterTextInputPlugin.o
4 warnings generated.
/Users/m/Projects/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:414:5: warning: Incorrect decrement of the reference count of an object that is not owned at this point by the caller [clang-analyzer-osx.cocoa.RetainCount]
    [array release];
    ^
/Users/m/Projects/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:1310:71: note: Calling 'start'
  NSUInteger currentIndex = ((FlutterTextPosition*)_selectedTextRange.start).index;
                                                                      ^~~~~
/Users/m/Projects/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:440:10: note: Calling 'positionWithIndex:'
  return [FlutterTextPosition positionWithIndex:self.range.location];
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/m/Projects/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:405:11: note: Calling 'initWithIndex:'
  return [[[FlutterTextPosition alloc] initWithIndex:index] autorelease];
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/m/Projects/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:410:7: note: 'self' is non-nil
  if (self) {
      ^~~~
/Users/m/Projects/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:410:3: note: Taking true branch
  if (self) {
  ^
/Users/m/Projects/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:413:22: note: NSArray literal is an object with a +0 retain count
    NSArray *array = @[@"test"];
                     ^~~~~~~~~~
/Users/m/Projects/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:414:5: note: Incorrect decrement of the reference count of an object that is not owned at this point by the caller
    [array release];
    ^~~~~~
/Users/m/Projects/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:506:25: warning: Nullable pointer is passed to a callee that requires a non-null 1st parameter [clang-analyzer-nullability.NullablePassedToNonnull]
  NSString* textAfter = [_textInputView
                        ^
/Users/m/Projects/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:507:19: note: Nullability 'nullable' is inferred
      textInRange:[_textInputView textRangeFromPosition:position
                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/m/Projects/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:506:25: note: Nullable pointer is passed to a callee that requires a non-null 1st parameter
  NSString* textAfter = [_textInputView
                        ^
/Users/m/Projects/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:740:3: warning: The '_markedTextStyle' ivar in 'FlutterTextInputView' was copied by a synthesized property but not released before '[super dealloc]' [clang-analyzer-osx.cocoa.Dealloc]
  [super dealloc];
  ^
/Users/m/Projects/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:544:3: note: Calling 'dealloc'
  [super dealloc];
  ^~~~~~~~~~~~~~~
/Users/m/Projects/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:740:3: note: The '_markedTextStyle' ivar in 'FlutterTextInputView' was copied by a synthesized property but not released before '[super dealloc]'
  [super dealloc];
  ^~~~~~~~~~~~~~~
/Users/m/Projects/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:740:3: warning: The '_textContentType' ivar in 'FlutterTextInputView' was copied by a synthesized property but not released before '[super dealloc]' [clang-analyzer-osx.cocoa.Dealloc]
  [super dealloc];
  ^
/Users/m/Projects/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:544:3: note: Calling 'dealloc'
  [super dealloc];
  ^~~~~~~~~~~~~~~
/Users/m/Projects/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:740:3: note: The '_textContentType' ivar in 'FlutterTextInputView' was copied by a synthesized property but not released before '[super dealloc]'
  [super dealloc];
  ^~~~~~~~~~~~~~~
```
On warnings, the process exits 0 so no issues are reported from `ci/lint.sh`.  

When I added `WarningsAsErrors: 'clang-analyzer-osx.cocoa.*'` to `.clang-tidy`, that didn't work because it looks like it's not actually being used passed into `clang-tidy` at the config is overridden as empty `config=`
https://github.com/flutter/engine/blob/1ae721c0230e8e3f3618f7e9d301ad2133b2fd61/tools/clang_tidy/lib/src/options.dart#L23

When I fix that and remove `config=` I start seeing additional clang tidy warnings like `google-`.
```
/Users/m/Projects/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:934:48: warning: statement should be inside braces [google-readability-braces-around-statements]
  if (range.location <= selectedRange.location)
```


So to sum up:
1. clang-tidy is only running on [`host_debug`](https://github.com/flutter/engine/blob/1ae721c0230e8e3f3618f7e9d301ad2133b2fd61/ci/lint.sh#L35), so not the iOS embedder. 
2. Warnings are not surfaced.  I'd like to turn on `WarningsAsErrors` for at `clang-analyzer-osx.cocoa.*`.
3. Google errors are not on at all since the `.clang-tidy` config file is not being passed into the `clang-tidy` command.

**Comment by zanderso** [comment]
We might have misunderstood the docs for the command line flags at https://clang.llvm.org/extra/clang-tidy/index.html
```
 --config=<string>              -
                                   Specifies a configuration in YAML/JSON format:
                                     -config="{Checks: '*',
                                               CheckOptions: [{key: x,
                                                               value: y}]}"
                                   When the value is empty, clang-tidy will
                                   attempt to find a file named .clang-tidy for
                                   each source file in its parent directories.
```

**Comment by jmagman** [comment]
> 3. Google errors are not on at all since the `.clang-tidy` config file is not being passed into the `clang-tidy` command.

Filed https://github.com/flutter/flutter/issues/93279 and https://github.com/flutter/flutter/issues/93280.

**Comment by jmagman** [comment]
https://flutter-review.googlesource.com/c/recipes/+/20260 will run `--lint-all` in CI on `ios_debug`.  https://github.com/flutter/engine/pull/29755 will turn `clang-analyzer-osx`.

**Comment by jmagman** [comment]
This is complete, `Mac clang-tidy` is linting `host_debug` and `ios_debug_sim`
https://ci.chromium.org/p/flutter/builders/prod/Mac%20clang-tidy/2066
